server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,433 @@
+        + "insert or update documents.";
+    private static final String UNSTABLE_MASTER_DEPLOYMENT_MANAGEMENT_IMPACT = "Scheduled tasks such as Watcher, ILM, and SLM will not "
+        + "work. The _cat APIs will not work.";
+    private static final String UNSTABLE_MASTER_BACKUP_IMPACT = "Snapshot and restore will not work.";
+
+    public static final Setting<TimeValue> VERY_RECENT_PAST_SETTING = Setting.timeSetting(
+        "health.master_history.very_recent_past",
+        DEFAULT_VERY_RECENT_PAST,
+        SMALLEST_ALLOWED_VERY_RECENT_PAST,
+        Setting.Property.NodeScope
+    );
+
+    public static final Setting<Integer> ACCEPTABLE_NULL_TRANSITIONS_SETTING = Setting.intSetting(
+        "health.master_history.acceptable_null_transitions",
+        DEFAULT_ACCEPTABLE_NULL_TRANSITIONS,
+        SMALLEST_ALLOWED_ACCEPTABLE_NULL_TRANSITIONS,
+        Setting.Property.NodeScope
+    );
+
+    public static final Setting<Integer> ACCEPTABLE_IDENTITY_CHANGES_SETTING = Setting.intSetting(
+        "health.master_history.acceptable_identity_changes",
+        DEFAULT_ACCEPTABLE_IDENTITY_CHANGES,
+        SMALLEST_ALLOWED_ACCEPTABLE_IDENTITY_CHANGES,
+        Setting.Property.NodeScope
+    );
+
+    public StableMasterHealthIndicatorService(ClusterService clusterService, MasterHistoryService masterHistoryService) {
+        this.clusterService = clusterService;
+        this.masterHistoryService = masterHistoryService;
+        this.veryRecentPast = VERY_RECENT_PAST_SETTING.get(clusterService.getSettings());
+        this.acceptableNullTransitions = ACCEPTABLE_NULL_TRANSITIONS_SETTING.get(clusterService.getSettings());
+        this.acceptableIdentityChanges = ACCEPTABLE_IDENTITY_CHANGES_SETTING.get(clusterService.getSettings());
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return CLUSTER_COORDINATION;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean explain) {
+        MasterHistory localMasterHistory = masterHistoryService.getLocalMasterHistory();
+        if (hasSeenMasterInVeryRecentPast()) {
+            return calculateWhenHaveSeenMasterRecently(localMasterHistory, explain);
+        } else {
+            return calculateWhenHaveNotSeenMasterRecently(localMasterHistory, explain);
+        }
+    }
+
+    /**
+     * Returns the health result for the case when we have seen a master recently (at some point in the last 30 seconds).
+     * @param localMasterHistory The master history as seen from the local machine
+     * @param explain Whether to calculate and include the details and user actions in the result
+     * @return The HealthIndicatorResult for the given localMasterHistory
+     */
+    private HealthIndicatorResult calculateWhenHaveSeenMasterRecently(MasterHistory localMasterHistory, boolean explain) {
```

Comment:
The name of this method indicates it's looking for the time something happened - the *when*. Which is not correct AFAICS. Should we name it differently? ie. `onSeenMaster` / `checkIfCurrentMasterIsStable` ? Or something along those lines?